Refactor Search Pipeline with Link Relevancy Assessment and Scraping Modes

NISH1001 opened the pull request:
**Title:** Refactor Search Pipeline with Link Relevancy Assessment and Scraping Modes

---
### Summary :memo:
This pull request refactors the `SearchPipeline` to provide more granular control over when web scraping is performed. It replaces the simple `enable_scraping` boolean with a `SearchPipelineScrapingMode` enum, allowing for modes where scraping is always on, always off, or conditional based on a new relevancy assessment step.

The core of this change is the integration of the `LinkRelevancyAssessor` tool, which scores search results *before* the expensive scraping step. This ensures that we only fetch full content for links that are highly relevant to the user's query, improving efficiency and the quality of the final output.

It's also a part of few major refactoring happening around `akd.agents.search.DeepLitSearchAgent` (see [this pr](https://github.com/NASA-IMPACT/accelerated-discovery/pull/203))

### Details
_Describe more what you did on changes._
1.  **Introduced `SearchPipelineScrapingMode` Enum**: Replaced the `enable_scraping: bool` field with `scraping_mode: SearchPipelineScrapingMode` in `SearchPipelineConfig`. The new modes are:
    * `ALWAYS_ON`: The previous default behavior.
    * `ALWAYS_OFF`: Disables scraping entirely.
    * `LINK_ASSESSMENT`: Uses an AI model to score link relevancy before deciding to scrape.
2.  **Integrated `LinkRelevancyAssessor`**: The `SearchPipeline` now orchestrates the `LinkRelevancyAssessor` to perform a batch assessment of all search results when in `LINK_ASSESSMENT` mode. The results of this assessment (score, and a `should_fetch_full_content` flag) are stored in the `SearchResultItem.extra` dictionary.
3.  **Refactored `SearchPipeline` Logic**: The pipeline's execution flow (`_arun` and `_process_single_result`) was updated to accommodate the new modes. It now determines whether to scrape based on the mode and the output of the relevancy assessor.
4.  **Standardized Metadata Storage**: Moved relevancy-related attributes like `should_fetch_full_content` and `relevancy_assessment` from direct class attributes to the `extra` dictionary on `SearchResultItem` for better consistency and flexibility.
5.  **Updated SearxNG Defaults**: Removed `google` from the default list of engines in `SearxNGSearchTool` to improve reliability.
6.  **Added Comprehensive Tests**: Added a new test suite for the `LinkRelevancyAssessor` and significantly updated the tests for `SearchPipeline` to cover all new scraping modes and logic.

### Bugfixes :bug:
- Fixed an issue where the domain context and reformulated query were not being correctly passed to the link assessor within the pipeline.
- Corrected the data structure access for relevancy scores and flags (using `result.extra.get(...)` instead of `result.attribute`) to prevent errors after the refactor.
- Ensured the final list of results returned by the pipeline correctly reflects the assessed and filtered items.

---

### Usage

#### Basic SearchPipeline Setup
```python
from akd.tools.search import (
    SearchPipeline,
    SearchPipelineConfig,
    SearchPipelineScrapingMode,
    SearxNGSearchTool,
    SearxNGSearchToolConfig,
    SearchToolInputSchema,
)
from akd.tools.scrapers.composite import CompositeScraper
from akd.tools.scrapers import (
    Crawl4AIWebScraper,
    SimpleWebScraper,
    SimplePDFScraper,
    DoclingScraper,
)

# Configure search tool
search_config = SearxNGSearchToolConfig(
    base_url="http://54.89.39.227:8080/",
    max_results=10,
    safe_search=2,
)
search_tool = SearxNGSearchTool(search_config, debug=True)

# Configure composite scraper
scraper = CompositeScraper(
    Crawl4AIWebScraper(),
    SimpleWebScraper(),
    SimplePDFScraper(),
    DoclingScraper(),
    debug=True,
)
```

#### Different Scraping Modes

**1. Always On Mode (Default behavior)**
```python
config = SearchPipelineConfig(
    debug=True,
    parallel_processing=True,
    max_concurrent_scrapes=5,
    scraping_mode=SearchPipelineScrapingMode.ALWAYS_ON,
)

pipeline = SearchPipeline(
    search_tool=search_tool,
    scraper=scraper,
    config=config,
    debug=True,
)
```

**2. Always Off Mode (No scraping)**
```python
config = SearchPipelineConfig(
    debug=True,
    parallel_processing=True,
    scraping_mode=SearchPipelineScrapingMode.ALWAYS_OFF,
)

pipeline = SearchPipeline(
    search_tool=search_tool,
    scraper=scraper,  # Still required but won't be used
    config=config,
    debug=True,
)
```

**3. Link Assessment Mode (Smart scraping)**
```python
config = SearchPipelineConfig(
    debug=True,
    parallel_processing=True,
    max_concurrent_scrapes=5,
    scraping_mode=SearchPipelineScrapingMode.LINK_ASSESSMENT,
)

pipeline = SearchPipeline(
    search_tool=search_tool,
    scraper=scraper,
    config=config,
    debug=True,
)
```

#### Configuring Link Relevancy Assessor

When using `LINK_ASSESSMENT` mode, you can customize the relevancy assessment behavior:

```python
from akd.tools.link_relevancy_assessor import (
    LinkRelevancyAssessorConfig,
    ScoringWeights,
)

# Custom scoring weights
custom_weights = ScoringWeights(
    topic_alignment_weight=0.30,        # Focus more on topic alignment
    content_depth_weight=0.20,         # Increased content depth importance
    evidence_quality_weight=0.25,      # Keep evidence quality high
    methodological_relevance_weight=0.10,  # Reduce methodological weight
    recency_relevance_weight=0.10,     # Keep recency moderate
    scope_relevance_weight=0.05,       # Reduce scope importance
)

# Custom assessor configuration
assessor_config = LinkRelevancyAssessorConfig(
    min_relevancy_score=0.4,           # Higher minimum threshold
    full_content_threshold=0.8,        # Higher threshold for full content
    assessment_batch_size=10,          # Process more in parallel
    enable_caching=True,
    debug=True,
    scoring_weights=custom_weights,
)

# Note: The pipeline will automatically create and configure the
# LinkRelevancyAssessor internally, but you can influence its behavior
# through the pipeline configuration or by subclassing SearchPipeline
```

#### Running the Pipeline

```python
# Execute search with pipeline
result = await pipeline.arun(
    SearchToolInputSchema(
        queries=["nepal landslide throughout the years"],
        max_results=5,
        category="science",
    ),
)

# Access results
print(f"Found {len(result.results)} results")

for search_result in result.results:
    print(f"Title: {search_result.title}")
    print(f"URL: {search_result.url}")
    print(f"Content: {search_result.content[:200]}")

    # Access relevancy metadata (when using LINK_ASSESSMENT mode)
    if "relevancy_score" in search_result.extra:
        print(f"Relevancy Score: {search_result.extra['relevancy_score']}")
        print(f"Should Fetch Full Content: {search_result.extra['should_fetch_full_content']}")

    print("---")
```

#### Accessing Relevancy Assessment Results

When using `LINK_ASSESSMENT` mode, each `SearchResultItem` will have additional metadata in the `extra` dictionary:

```python
for result in pipeline_output.results:
    # Standard fields
    print(f"Title: {result.title}")
    print(f"URL: {result.url}")

    # Relevancy assessment metadata (available in LINK_ASSESSMENT mode)
    relevancy_score = result.extra.get("relevancy_score", None)
    should_fetch = result.extra.get("should_fetch_full_content", False)
    assessment_details = result.extra.get("relevancy_assessment", {})

    print(f"Relevancy Score: {relevancy_score}")
    print(f"Will fetch full content: {should_fetch}")

    if assessment_details:
        print(f"Topic Alignment: {assessment_details.get('topic_alignment')}")
        print(f"Evidence Quality: {assessment_details.get('evidence_quality')}")
        print(f"Overall Relevance: {assessment_details.get('overall_relevance')}")
```

---

### Checks
- [x] Tested Changes (100% pass, 76% coverage)
- [ ] Stakeholder Approval

NISH1001 committed:
Add SearchPipelineScrapingMode enum

NISH1001 committed:
Add link relevancy assessor to search pipeline

- Dependency-inject `link_relevancy_assessor` to `SearchPipeline`
- Refactor `LinkRelevancyAssessor` class to conform to SearchResultItem
  structure. Move all the non-existent keys to `extra`

NISH1001 committed:
Add SearchPipelineScrapingMode enum to akd.tools.search.__init__

NISH1001 committed:
Bugfix search tests for new refactor

NISH1001 committed:
Add tests for link relevany assessor

- Add `tests/tools/test_link_relevancy_assessor.py`

NISH1001 committed:
Remove obsolete comment

NISH1001 committed:
Refactor scraping checks in search pipeline

NISH1001 committed:
Bugfix domain context and reformulated query within search pipeline

- Assess the params via kwargs

NISH1001 committed:
Make search tool optoinal in SearchPipeline

- Default to `SearxngSearchTool`

NISH1001 committed:
Bugfix searh result list from search pipeline

- Return `filtered_results` instead of `assessed_results`

NISH1001 committed:
Remove google from default engine in searxng

- Also add `safe_search` param to searxng config and default to 2
  (strict)

NISH1001 committed:
Merge branch 'develop' of github.com:NASA-IMPACT/accelerated-discovery into enhance/search-pipeline

NISH1001 committed:
Merge branch 'develop' of github.com:NASA-IMPACT/accelerated-discovery into enhance/search-pipeline

muthukumaranR commented on akd/tools/search/pipeline.py:
```diff
@@ -134,23 +131,55 @@ def __init__(
         if debug:
             logger.debug("Initialized SearchPipeline with:")
             logger.debug(f"  - Search tool: {search_tool.__class__.__name__}")
             logger.debug(f"  - Resolver: {self.resolver.__class__.__name__}")
```
```suggestion
            logger.debug(f"  - Search tool: {self.search_tool.__class__.__name__}")
            logger.debug(f"  - Resolver: {self.resolver.__class__.__name__}")
```

muthukumaranR replied:
in case search_tool is not passed

muthukumaranR commented on akd/tools/link_relevancy_assessor.py:
```diff
@@ -347,25 +347,25 @@ async def _assess_query_alignment(
                             "best_query": "reformulated",
-                            "original_score": orig_result.relevancy_score,
-                            "reformulated_score": reform_result.relevancy_score,
+                            "original_score": orig_result.score,
+                            "reformulated_score": reform_result.score,
```
if the intent is to capture the original score, it's being replaced by the reformulated score so you are accessing the reformulated score here

NISH1001 replied:
oh ya. this is a bug. thanks for catching that.

NISH1001 replied:
Fixed.

NISH1001 committed:
Update akd/tools/search/pipeline.py

Co-authored-by: muthukumaran R <mr0051@uah.edu>

NISH1001 committed:
Reset memory for each link asessment in link assessor

- By default the `instructorbaseagent` tracks conversation history
- TODO: Make the base agent stateless by default

NISH1001 committed:
Bugfix reformulated score assignment in link assessor
